src/node_crypto.cc

```diff
@@ -155,7 +155,6 @@ std::string extra_root_certs_file;  // NOLINT(runtime/string)
 X509_STORE* root_cert_store;
 
 // Just to generate static methods
-template class SSLWrap<TLSWrap>;
```

Comment:
> Actually, I’m a bit confused. 😄

Welcome to my world 😄 . 
I think I did a bad job of trying to explain what I think is going on. Let me try again.



> For a linker error to be generated, the compiler would have first needed to be told that there are explicit instantiations generated elsewhere; but I don’t see anything like that?

Lets focus on the removal of the explicit class instantiations which this PR is about:
```c++
template class SSLWrap<TLSWrap>;
```
And the explicit instantiations of its protected member functions:
```c++
template void SSLWrap<TLSWrap>::DestroySSL();
template void SSLWrap<TLSWrap>::WaitForCertCb(CertCb cb, void* arg);
template void SSLWrap<TLSWrap>::SetSNIContext(SecureContext* sc);
template int SSLWrap<TLSWrap>::SetCACerts(SecureContext* sc);
```

Lets say we want to also remove the explicit template instantiations of the protected members as well. 

Lets comment out DestroySSL which is used in [tls_wrap.cc](https://github.com/nodejs/node/blob/d54ec726cc9d0d293e61dc0dba61a09429e4cbaf/src/tls_wrap.cc#L829):
```c++
TLSWrap* wrap;
...
wrap->SSLWrap<TLSWrap>::DestroySSL();
```
If we take a look at the unresolved symbols for `tls_wrap.o` we can find the following:
```console
$ nm -gU out/Debug/obj.target/node/src/tls_wrap.o|grep DestroySSL
00000000000048f0 T __ZN4node7TLSWrap10DestroySSLERKN2v820FunctionCallbackInfoINS1_5ValueEEE

$ c++filt __ZN4node7TLSWrap10DestroySSLERKN2v820FunctionCallbackInfoINS1_5ValueEEE
node::TLSWrap::DestroySSL(v8::FunctionCallbackInfo<v8::Value> const&)
```
My understanding is that the compiler will check the types of the template when compiling tls_wrap.cc, but then just create a symbol to be resolve later by the linker. 

If we take a look at the symbols created for `DestroySSL` in `node_crypto.o` we find:
```console
$ nm -g out/Debug/obj.target/node/src/node_crypto.o | grep DestroySSL
0000000000009250 T __ZN4node6crypto7SSLWrapINS0_10ConnectionEE10DestroySSLEv

$ c++filt __ZN4node6crypto7SSLWrapINS0_10ConnectionEE10DestroySSLEv
node::crypto::SSLWrap<node::crypto::Connection>::DestroySSL()
```
In this case the compiler only instantiated (generated the function specialised for type node::crypto::Connection) which was done implicitly/automatic as it is [used](https://github.com/nodejs/node/blob/4c847347c3e8dd3198be2d2d71728521a1f42281/src/node_crypto.cc#L2761) in node_crypto.cc. But there is no such instantiation of DestroySSL for a type of TLSWrap as there is no usage of it (it is used in a separate compilation unit) which will lead to a link error. 

If we comment back the explicit instantiation of DestroySSL we can inspect `node_crypto.o` once again and see that it generated the function specialised for TLSWrap:
```console
$ nm -g out/Debug/obj.target/node/src/node_crypto.o | grep DestroySSL
0000000000009250 T __ZN4node6crypto7SSLWrapINS0_10ConnectionEE10DestroySSLEv
000000000001bbb0 T __ZN4node6crypto7SSLWrapINS_7TLSWrapEE10DestroySSLEv

$ c++filt __ZN4node6crypto7SSLWrapINS_7TLSWrapEE10DestroySSLEv
node::crypto::SSLWrap<node::TLSWrap>::DestroySSL()
```
Now, this is my understanding of what is going on so please correct me if I'm wrong about anything here. 

@addaleax Does that make sense?